mistralrs-core/src/sampler.rs

```diff
@@ -488,7 +488,7 @@ impl Sampler {
             let match_indices = toks
                 .par_iter()
                 .enumerate()
-                .take(toks.len() - 1)
+                .take(toks.len().saturating_sub(1))
```

Comment:
Just ran into this problem before I noticed your commit. But why can `toks.len()` be 0 here in the first place? That doesn't make sense to me. If I enter a prompt (e.g. "test"), then `apply_dry_penalty` is called with an empty context. Why?

Reply:
So this can only be caused when sampling the token resulting from prompt processing. We discard all prompt tokens from the penalty context:

> https://github.com/EricLBuehler/mistral.rs/blob/master/mistralrs-core/src/pipeline/sampling.rs#L290

Does DRY require those tokens to be included?

Reply:
I'm afraid I don't understand. Why is sampling needed during prompt processing?

Sure, the transformer will by construction always output logits for each token position. But samplers are only getting involved once a token is actually drawn from the resulting distribution. And that only happens when new tokens are generated, right? In which case the context should always be non-empty after a prompt has been entered.

Reply:
> Why is sampling needed during prompt processing?

After the model processes the prompt, it produces a token distribution as a result and we sample that. The `Sampler::apply_penalties` method has inherent support for the case where the generated tokens do not exist yet, as we just iterate over that context. Perhaps we need a case here to handle this?

Reply:
I feel bad for taking so much of your time with this, but I just don't get it. Why does sampling ever happen with an empty context? The only purpose of sampling (i.e., *drawing* from the probability distribution, rather than merely generating the distribution) is to generate a new token, no?

Let's say the user enters the prompt "Hello", and then runs generation. Why is `Sampler::apply_penalties` called with an empty context, rather than with context "Hello"? The program doesn't need to sample during prompt processing (since no tokens are being generated at previous positions), so why is this happening?

Reply:
Thank you, I understand now. So this is intended to be a feature?

It seems that this breaks many assumptions made by frontends (and users) about how things work. The most problematic effect is that it makes sampling a stateful affair, where the original prompt range is somehow "remembered" during generation, and the context given to the samplers doesn't match the input given to the transformer.

In a chat interface, this means repetition penalties cannot take into account previous messages while generating a new one... which is the exact opposite of what we want, since models tend to repeat previous messages. But it also seems wrong philosophically. If you cancel the generation process midway, and then restart it from that position, you might get *different* tokens than you would have if you had allowed it to complete, because now the prompt includes the partially generated output.

I don't believe other loaders do this, and IMO this mechanism should indeed be removed completely, for all samplers.

Reply:
Thanks for pointing this out, I agree that the statefulness makes this incorrect!

> I don't believe other loaders do this, and IMO this mechanism should indeed be removed completely, for all samplers.

Sounds good! I've removed this functionality completely now.